### PR TITLE
Add debugging routines

### PR DESCRIPTION
**Pull request by nmatthews-witekio**
Add debugging routines

Macro definition with (()) allows all debugging strings to be removed from the executable if required -- keeping executable size smaller -- but allowing debugging to be turned on if/when required.

### CODE REVIEW

**Comment by nmatthews-witekio** [comment]
Clarification:

Think of a debug routine like printf...

one_bracket_debug("This is the value %d", x);
then the one_bracket_debug macro has 2 arguments -- "This is the value %d" and x
This means you have to use __va_args__ to deal with the extra multiple optional arguments.

if you use the double-bracket technique then:
two_bracket_debug(("This is the value %d", x));
then the two_bracket_debug macro has 1 argument -- ("This is the value %d", x)
So it can work without needing __va_args__i.e. on an older compiler.

Also if you do:
#define two_bracket_debug(x)
the compiler knows that the argument x is unused -- so it can throw away all of -- ("This is the value %d", x)
so the string literal "This is the value %d" probably doesn't make it into the executable

You could use __va_args__:
#define one_bracket_debug(x...)
but there's probably a bit less chance for the compiler to be able to discard the string "This is the value %d"

That's all.
It looks a bit weird in the code -- but actually it sort of makes sense -- especially if you do
#define two_bracket_debug(x) printf x
then the single argument -- ("This is the value %d", x) -- gets rewritten easily as
printf ("This is the value %d", x)

**Comment by nmatthews-witekio** [comment]
If you use a #define that just discards its argument -- then the pre-processor stage will just remove the  argument -- so the compilation stage won't even "see" the argument -- as it'll have already been removed.

**Comment by nmatthews-witekio** [comment]
This still needs reviewing @PatriciaFieldhouse.

**Review by nik-markovic — CHANGES_REQUESTED** [review]
I can see why we would want a mechanism to control deblug/log levels - some sort of logger. That way we will not be spamming the console if the user does not want that.

I do not however see why we would want a compile time define to turn all logging on or off completely - including errors, which are crucial. It's not like this is an embedded system and we need to save 10k worth of .text/.rodata section strings, or we should not use printf, or we need to optimize for speed here. If this is supposed to be a prelude to a future change, I would welcome it, but in the current form it looks a bit odd. The compile time define should not be bound to DEBUG, but rather some IOTC prefixed define that defines a log level and defaults to INFO by default.

Why not just improve this right away and have a runtime logger or a set of cacros that support log levels, if that was what you wanted to achieve? I'm sure there are some logger open source logging facilities for C that we can pick up or do a log lecel based macro implementation.

I saw your comment about double braces and compiler optiomization. There is no rason for any modern compiler not to optimize the macro results even if it's declared like this without the odd extra brackets
```
nmarkovi@u20:~$ cat t.c
#include <stdio.h>

#ifdef IOTC_LOGGING
#define LOGME(...) printf(__VA_ARGS__)
#else
#define LOGME(...)
#endif 

int main() {
  LOGME("hello %s", "there\n");
}

nmarkovi@u20:~$ strings a.out  | grep hello
nmarkovi@u20:~$ gcc t.c
nmarkovi@u20:~$ ./a.out 
nmarkovi@u20:~$ strings a.out  | grep hello
nmarkovi@u20:~$ gcc -DIOTC_LOGGING t.c
nmarkovi@u20:~$ ./a.out 
hello there
nmarkovi@u20:~$ strings a.out  | grep hello
hello %s
```

In a nutshell, I see room for improvement that can be addressed by adding some logging functionality. For example, \n should not be a thing you add to your log statements in some future impelementation. However, I am not not this low level approach with extra braces and whatnots. We can talk more about this offline if you would prefer. This can probably go in as a preliminary logging framework but without the braces.

**Comment by nmatthews-witekio** [comment]
Not sure what happened -- have adjusted an opened https://github.com/avnet-iotconnect/iotc-generic-c-sdk/pull/46
